### ISSUE TITLE
IBM 1620: ^E doesn't interrupt I/O... WEDGE likely

### ISSUE THREAD

**Issue by DaveWise2** [simh/simh#436]
- #### Context

This applies to all I/O, but it's only a show-stopper for typewriter input.

1620 console typewriter input is not character-by-character.  It is not even line-by-line.  Carriage Return is ignored.  It is not even a character.  The CPU accepts input indefinitely, until the operator manually ends it by pressing the RELEASE console button.  Then the operator presses START to resume execution.

For convenience, the typewriter has a "RELEASE-START" key.  SIMH treats the Enter key as RELEASE-START, but sometimes it is essential that the CPU stay halted after I/O is done.

You see, Backspace/Delete/Rubout don't exist either.  If you mis-type, you're simply screwed.  Therefore, virtually all programs that accept typewriter input incorporate a loop whereby the operator can make the program retry.  The convention is to set Console Switch 4 (sw4), hit START (the program sees sw4 and repeats the input operation), retype, RELEASE, turn off sw4, and START.  But SIMH doesn't give you a chance to turn off sw4, and you're trapped in an infinite loop of retries.

I was hoping that I could type without hitting Enter, then hit ^E (WRU) to get SIMH's attention.  Then I'd 'd sw4 0' and 'go'.  But when I hit ^E, nothing happens.  I bet SIMH only checks it between instructions.  But it needs to be RELEASE, exiting the I/O without resuming.

If ^E is hard, how about semicolon?  It's a printing character but outside the 1620 character set, so maybe it can be processed more easily, by whoever handles the RNTY/RATY (Read Numeric/Alphameric TYpewriter) opcodes.

I'm unfamiliar with SIMH internals so pardon me if these proposals are "not even wrong".  Please suggest places in the code that I should study.

- #### the output of "sim> SHOW VERSION" while running the simulator which is having the issue
sim> show ver
IBM 1620 simulator V4.0-0 Beta
        Simulator Framework Capabilities:
                32b data
                32b addresses
                no Ethernet
                Idle/Throttling support is available
                Virtual Hard Disk (VHD) support
                RAW disk and CD/DVD ROM support
                FrontPanel API Version 2
        Host Platform:
                Compiler: Microsoft Visual C++ 15.00.30729.01
                Simulator Compiled as C arch: x86 (Release Build) on Jan  2 2017 at 14:08:49
                Memory Access: Little Endian
                Memory Pointer Size: 32 bits
                Large File (>2GB) support
                SDL Video support: No Video Support
                PCRE RegEx support for EXPECT commands
                OS clock resolution: 1ms
                Time taken by msleep(1): 1ms
                OS: Microsoft Windows [Version 6.1.7601]
                Architecture: x86 on AMD64, Processors: 36
                Processor Id: Intel64 Family 6 Model 63 Stepping 2, GenuineIntel, Level: 6, Revision: 3f02
        git commit id: d77e3e19
sim>
- #### how you built the simulator or that you're using prebuilt binaries

Prebuilt binary.

- #### the simulator configuration file (or commands) which were used when the problem occurred.

No config file.

- #### the expected behavior and the actual behavior

Expected:

[program didn't like input, emits message and halts]
sim>d sw4 1
sim>go
\> [program waiting for input]
\>       100 FORMAT (I3,F8.4)  ^E [breaks to SIMH prompt]
sim>d sw4 0
sim>go
[program resumes, using corrected line]

Actual:

[program didn't like input, emits message and halts]
sim>d sw4 1
sim>go
\> [1620 program waiting for input]
\>       100 FORMAT (I3,F8.4)  ^E [nothing happens]
[shrug and hit Enter]
\> [1620 program waiting for input again]
[can't clear sw4... kill SIMH and start over, hoping for no typo's]

- #### you may also need to provide specific pointers to data files that may be necessary to demonstrate the problem

Contact me.  I was linting a source tape with the FORTRAN Pre-Compiler (part number FO-005), but any program with a "goof loop" will do.

It's particularly evil with FO-005, because sw4 is not only the goof signal, but also one of the normal operating modes!  See C26-5619-4_1620_Fortran_Mar64.pdf , page 78, "Program Switch Settings".

Thanks,
Dave Wise
Keeper of the last running 1620

**Comment by DaveWise2** [comment]
I looked over the code and tried out a mod.
It seems to work, but it's inelegant, because it will lose if applied to other cases.
I'm not happy about that.

Just add two lines at the beginning of tti_read() in i1620_tty.c .

```
do {
    t = sim_poll_kbd ();                                /* get character */
    if (stop_cpu == 1)         /* NEW */
        return SCPE_STOP;      /* NEW */
    } while ((t == SCPE_OK) || (t & SCPE_BREAK));       /* ignore break */
```

SIMH works differently than the real machine.  The latter fetches (which advances the PC), then executes, so a RELEASEd I/O instruction is over and done.  OTOH, SIMH doesn't advance until the instruction is finished, so if you go or step after breaking, the instruction is repeated.  This is okay for the goof loop, but not for output, paper tape, card, etc.  I will think about it.

**Comment by DaveWise2** [comment]
Another two lines, added to i1620_cpu.c .

```
    case OP_RA:
    case OP_WA:
        if ((PAR & 1) == 0) {                           /* P even? */
            reason = STOP_INVEAD;                       /* stop */
            break;
            }
    case OP_DN:
    case OP_RN:
    case OP_WN:
        dev = get_2d (ADDR_A (saved_PC, I_IO));         /* get IO dev */
        f0 = M[ADDR_A (saved_PC, I_CTL)] & DIGIT;       /* get function */
        f1 = M[ADDR_A (saved_PC, I_CTL + 1)] & DIGIT;
        if ((dev < 0) || (iodisp[dev] == NULL))         /* undefined dev? */
            reason = STOP_INVIO;                        /* stop */
        else reason = iodisp[dev] (op, PAR, f0, f1);    /* call device */
/*
/* If we break out of an I/O operation via ^E, the SIMH analog of
 * the 1620 RELEASE button, it's not an error, and we should not
 * re-enter the instruction on Step or Go.
 */
        if (reason == SCPE_STOP)                        /* RELEASE i.e. ^E */
            saved_PC = PC;                              /* commit inst */
        break;
```

That seems to do it for RxTY.  Now you type the corrected line, hit RELEASE, uh, ^E, clear sw4, step 2, set sw4, and go merrily along.  Just like the real thing.  I will review the code to make sure iodisp can't return SCPE_STOP for some other reason.

RELEASE ends all I/O on a real 1620, but the other Rxxx cases are not interesting for SIMH.

I will try to implement RELEASE for WNTY and DNTY so you can abort spew.  They also need throttling, but that's different story.

**Comment by markpizz** [comment]
There was a change in the 4.x code a while back that made calling sim_poll_kbd() behave consistently across platforms.  Specifically, on *nix platforms (Linux & OS X), WRU detection is done with a signal that sets stop_cpu.  On those platforms, simulators noticed stop_cpu independently of any call sim_poll_kbd().  The simulator that inspired this change, and potentially other simulators (which were written on *nix platforms), have no coherent code path in their console device code to handle sim_poll_kbd() returning SCPE_STOP.  The 4.x sim_console code now sets stop_cpu when sim_poll_kbd() sees a WRU character and it DOES NOT return SCPE_OK.  This 1) requires some device to periodically call sim_poll_kbd() so that WRU characters will be noticed on Windows host platforms, and 2) allows consistent control flow in the simulator to detect stop_cpu being set at the right place so that execution can be stop and properly resumed after a WRU has been entered.  It seems that with the 3.X code, a user entering WRU will have no effect on a I1620 simulator on a non-Windows host.

Meanwhile, back to the problem that @DaveWise2  reported...  It seems that the calling sim_poll_kbd() directly from within the instruction execution path (as opposed from within an UNIT's service routine), at best makes consistently stopping a simulator hard to do.

It would seem to me that i1620_tty.c (tti_rnum(), tti_ralp() and tti_svc()) should be modified to keep track of whether a read is actually in progress and if it is, then accumulate and validate any characters returned from sim_poll_kbd() into the a buffer and complete it when \r is seen.  If stop_cpu is noticed while in tti_svc() return SCPE_STOP should happen.  When a read completes, saved_PC should be advanced to point at the next instruction.  This approach will probably provide useful behavior if you attempt to throttle execution as well.

**Comment by DaveWise2** [comment]
I don't think I'm understanding what you say about "read is actually in progress", Mark.  It seems to me that that _is_ what's being done: read is in progress because tti_rnum() or tti_ralp() was called, and it will keep getting called, once per char, until it's over, i.e. char == \r.

I removed my change
`if (stop_cpu == 1) return SCPE_STOP;`
from tti_read() and moved it to tty_svc().
^E did not interrupt RxTY.  I didn't expect it to work, because the calls to tti_ralp() and tti_rnum() are inside small "almost forever" for loops, so tty_svc() never gets called during RxTY.

**Comment by markpizz** [comment]
My comments were mostly for Bob Supnik (the original author of the whole simh system and this simulator in particular), or whomever actually fixes the problem(s) you're pointing out.

**Comment by DaveWise2** [comment]
Thanks, Mark.  I don't know how to behave.  Where is the project policy?

**Comment by markpizz** [comment]
Dave,

There is no behavior problem on your part here.  This device's behavior is somewhat rare in the existing set of simulators and as such the model chosen to implement it hasn't had a lot of scrutiny.  My comments were some ideas about how the problem could be solved more generically and still fit in well with the other functional simulator framework features (^E escape and throttling for example).  Unless you've written several simulators and/or simulator devices, how to structure device behavior to achieve these results may not be obvious.  My comments suggest that the structure chosen to implement this device doesn't match well with the normal expectations of the simulation framework and a restructuring is the right approach to get things working as opposed to trying to make minor changes to the existing implementation.

- Mark

**Comment by DaveWise2** [comment]
I wondered about that.  I am eager to see what Bob et. al. have to say.  Realistic throttling would require huge modifications.  Right now I1620 simulated instructions all take exactly one clock tick, if I understand correctly.  The real 1620 varies wildly by opcode and data, with a Model I CPU taking anywhere from 160 microseconds (eight 20us machine cycles) to hundreds of thousands.  I/O is ten chars/second for the console typewriter and 150 for the paper tape reader, but I doubt anybody really wants it to go that slow.

**Comment by archivisth** [comment]
I'm still investigating this issue, but I have a couple of initial observations.

The 1620 Mark II used a Selectric for its typewriter. It<does> have a backspace character, and it behaves pretty much like you think it should: it backspaces the carriage, backspaces the memory pointer that RN or RA is using, and continues to accept input. Unfortunately, it is not implemented. ;) But this should be part of any fix.

The manual mode of using the typewriter is not supported. The equivalent of INSERT-type-R/S is to halt the machine and use standard SimH commands:

sim> id 0-100
0:   digit
1:   digit
etc
sim> run 0

This may not provide an authentic recreation of the "type in the bootstrap sequence", but it works. The simulator supports BOOT PTR and BOOT CDR as well. I have not found a standard disk bootstrap, but that could easily be added.

When the 1620 is executing an RN or RA, it remains in "automatic" mode - the manual says so. I'm not exactly sure what this implies, but the normal sequence for input is RN/RA-type something-hit R/S. The return key on the typewriter does not generate a valid keycode, so using it as the R/S key seems to be good enough. It would be easy enough to allow this to be changed to a different keycode.

You wrote:

> SIMH works differently than the real machine. The latter fetches (which advances the PC), then executes, so a RELEASEd I/O instruction is over and done. OTOH, SIMH doesn't advance until the instruction is finished, so if you go or step after breaking, the instruction is repeated. This is okay for the goof loop, but not for output, paper tape, card, etc. I will think about it.

You are correct that SimH only commits when an instruction has finished successfully. This was done so that fatal stops, like invalid characters and invalid addresses, pointed at the failing instruction, rather than the next one. It seems to be incorrect behavior for IO and will have to be fixed.

Therefore, what I have so far is:

1. Add backspace capability for the Model II typewriter.
2. Figure out why ^E fails to interrupt RA/RN (it's supposed to).
3. Commit PC updates before instruction execution, like the real machine (and deal with any fatal error cases where that shouldn't happen).

I'm going to consult Dave McBride, who has deep expertise in the 1620 hardware, before proceeding.

**Comment by DaveWise2** [comment]
Right, Bob.  Glad to hear from you!  Note that backspace is not supported in Model I.

However, I implemented it back in 1979 on my own machine, adding wires and using up almost every spare gate.  I activate it by typing not backspace (there is no key contact) but the Cyrillic "ZHA" character, which IBM used - for output only - to represent a bit pattern with good parity but no translation.  I overload that function with "printing rubout" on input.  I added schematic pages to the binder, and you can barely tell they're new.

In 2000 - when I was helping the Computer History Museum restore their 1620 by designing and building a hardware peripheral emulator - I had the honor of discussing it with the late Joe Crespo, an IBM engineer from the San Jose group that brought up the prototype in 1958.  He told me that they had considered backspace, and would have implemented it almost exactly the way I did!  But decided not to since it required the optional Indirect Addressing feature.

I would like to participate in the discussion with Dave M, since I too know the hardware well.  (Model I only.)

Would you like to review my changes that make tty I/O interruptible?
I simulate RELEASE with ^E.  This allows one to do things between RELEASE and START.
To get out of the goof loop, you type your corrected line, hit ^E, turn off sw4, and go.  (In the case of the Fortran Pre-Compiler, 'sw4 on' is one of the normal operating modes besides its goof function, so I step past the BC4 instruction, turn sw4 back on, then go.)

**Comment by DaveWise2** [comment]
Automatic mode is "not halted".  But it doesn't necessarily mean the CPU is churning through instructions.  When it executes an I/O instruction, the clock stops, and machine cycles only happen when the peripheral device clocks the CPU with or for another character.  Nothing else goes on until the I/O is finished.

If you don't count the RESET button or turning off the power, there are only two ways of ending I/O.  One is for the device to signal that it's over.  This case can happen with input devices like the paper tape reader, the card reader, and the disk drive.  For paper tape, it's when it reads an EOL character.  For card, it's when the reader has transferred 80 characters.  For output, it's over when the data says it's over.  For WAPT/WNPT, it's when the CPU transfers a record mark.  On paper tape, that punches an EOL.  For WxCD, it's 80 characters.  For DNxx (Dump Numerical), it's when the CPU wraps around core.  A one-instruction core dump, but programs use it all the time to punch a record mark **as a record mark** instead of as an EOL.

The other escape from I/O is when the operator hits the RELEASE button.  This is the only way out of RxTY.  As a convenience, the typewriter is also equipped with a special "RELEASE-START" key, which, like SIMH, simulates that two-button sequence, which is fine as long as you don't have to flip any switches etc.  But the canonical way out of typewriter input is RELEASE, <maybe do some stuff>, START.

**Comment by DaveWise2** [comment]
I'd like to emphasize just how prehistoric it feels operating this machine.  From 1959 until 1963 (levels B and C), there was no disk drive and no OS.  In between programs, the CPU sat halted, in "Manual mode".  You threaded a paper tape or put a card deck in the hopper, and started the machine.  Programs ended with a HALT instruction, and each program was a boot.  This has implications for SIMH remote operation, which IIUC assumes the target is always spinning in a loop checking for I/O.

After Monitor I appeared, disk-equipped (D-level and up) 1620's running it would be in Automatic mode most of the time, but usually sitting paused inside the RATY instruction waiting for operator commands.  The RATY didn't finish until the operator hit the RELEASE button or the typewriter RELEASE-START key.  I don't know what programs did when they finished.  Maybe still HALTed, and the op would have to do something to get back into the monitor.  Or maybe the new compiler would insert a jump back to the monitor.  But either way, the real 1620 not only does not interrupt on I/O, it doesn't even poll it.

**Comment by TomMcBride** [comment]
As I see it, many, many things are working well with i1620. Entering cntrl+E stops simulation at the end of the current instruction and displays the next instruction to be executed. When stepping through a program, each instruction is displayed before it’s executed. When an instruction has an error such as a bad address, an error is issued, simulation stops and the failing instruction is displayed. This all seems perfect to me. The problem here is that if the current instruction is a read from the typewriter, cntrl+E does not break in and stop the simulation. This is a common need but an unusual situation for the simulator since it would mean stopping in the middle of executing an instruction. Nonetheless, I think this special case should be handled. Somehow the cntrl+E must be recognized during a typewriter read and the simulation stopped … effectively stopping the read in midstream. Whatever character(s) have already been entered into storage should be left there in the same way they would have been on a real machine doing a read if the release key were pressed. I would say that the prematurely terminated read should be treated as if it completed normally, simulation stopped and the next instruction following the read should be displayed. When GO is typed, that displayed instruction should be the next one executed. There should be no error indicated on the read. Stopping the simulation makes it possible to do anything such as changing a switch, displaying storage, etc.

It is important for the simulator to act like a real machine by placing each character into storage as it is entered. This has bearing on the other completely independent issue mentioned: that of simulating the correction feature of the Model 2 for typing errors. I never actually used a Model 2 machine so I cannot report details of this feature from personal experience. However, it is incompletely specified in the 1620 Model 2 manual. The machine uses a register to track where the next character typed is to be placed into storage and increments it by one (or two for alphanumeric) for each character typed. Each time the correction key is pressed, the register is decremented. Decrementing the register does not remove the incorrect character(s) from storage and I see nothing to protect the program from having the operator correct backwards more characters that he has typed which would allow typing into data areas in front of the planned input area. That would clearly be undesirable. After a read from the typewriter, there is no way for the program to determine exactly how many characters were typed (aside from filling the input area with some known character and checking to see how many of them were altered on the input operation). For example, if an operator were supposed to enter a name like TOM but mistakenly typed TOOM, then pressed the correct key twice, then pressed M, the data in storage would be TOMM and the program would not be able to tell that the operator intended to enter only three characters.

**Comment by DaveWise2** [comment]
The Model I backspace mod that I conceived in 1979 works exactly as you describe; it only decrements OR2, without attempting to restore previous memory content or avoid address wraparound.

What do you think of my changes in the second and third comments at the start of the thread?

Finding the end of typewriter input has always been a problem.  Programs that expect a record mark don't suffer from TOMM, but they are a minority; most just preset ram to a non-character, read, and search backward.  And that's why backspace, although it's convenient, can't obviate the Goof Loop.

Recently - after 38 years - I realized that Indirect Addressing (which includes Decrement-MAR-2 logic) is actually orthogonal to backspace; I could have simply decremented by 1 twice.  I wish Joe were here, so I could see the look on his face.

**Comment by TomMcBride** [comment]
Dave,
I added your two changes to my code and it seems to work OK. I must defer to Bob and Mark for a decision as to whether this is the best way to accomplish the function and to make sure there are no other problems with this method. I tried it on my Mac and have copied the console file below. It's a bit clumsy to manage the correction but is workable.

Console log:

Thomass-MBP:development Tom$ sps bug
476 lines written
Begin assembly of IBMSPS program bug ...

IBM 1620 simulator V4.0-0 Beta        git commit id: 7317645d
Logging to file "tmp.con"
CDP: creating new file

SPS PROCESSOR  FOR 1620/1710 CARD I/O SYSTEM,DATED 1/1/1962

HALT instruction, PC: 12 (B    17518,-67901,27810)

END OF PASSI  
HALT instruction, PC: 12052 (B    01792,03100)

     RETRY  RCTY  
        00402 34 00000 00102
            WATY PROMPT
        00414 39 00487 00100
            RATY X
        00426 37 00541 00100
            BC4  RETRY
        00438 46 00402 00400
            RCTY  
        00450 34 00000 00102
            WATY GOOD
        00462 39 00521 00100
            B    RETRY
        00474 49 00402 00000
     PROMPT DAC  17,ENTER SOMETHING-@
        00487 00017
     GOOD   DAC  10,GOOD READ@
        00521 00010
     X      DAS  100
        00541 00100
            DEND RETRY
        00402
END OF PASSII 
00402   RETRY   00487 *PROMPT   00521   GOOD    00541   X       
HALT instruction, PC: 12676 (BC3  01792)
Goodbye
Log file closed
Thomass-MBP:development Tom$ run bug
33 lines written
Begin execution of program bug ...
Output will be in tmp.con for console, tmp.pun for cards punched, tmp.lpt for lines printed.

IBM 1620 simulator V4.0-0 Beta        git commit id: 7317645d
Logging to file "tmp.con"
CDP: creating new file
LPT: creating new file

HALT instruction, PC: 12 (B    00402,-67901,27810)

ENTER SOMETHING->oops this is wrong
Simulation stopped, PC: 438 (BC4  00402)
sim> de sw4 1
sim> st

Step expired, PC: 402 (RCTY)
sim> st


Step expired, PC: 414 (WATY 00487)
sim> st
ENTER SOMETHING-
Step expired, PC: 426 (RATY 00541)
sim> st
>this is correct

Step expired, PC: 438 (BC4  00402)
sim> de sw4 0
sim> go

GOOD READ
ENTER SOMETHING->
Simulation stopped, PC: 438 (BC4  00402)
sim> quit
Goodbye
Log file closed
Thomass-MBP:development Tom$

**Comment by DaveWise2** [comment]
As I tried to get IBM's MI-015 HASH TOTAL program running under SIMH (see comments in https://github.com/simh/simh/issues/446), I realized that my idea here to treat ^E as RELEASE is fine for tty but not extensible to paper tape.  At last I think I understand what Mark suggested.  SIMH needs to be able to break to command prompt from an in-progress I/O instruction, then either RELEASE out of the instruction, or resume back into it where it left off.

It's a common event in a physical 1620 shop to forget to load the tape.  You hit START, and (a) the AUTOMATIC and MANUAL lamps light up simultaneously to indicate a paused state similar to hitting SCE (Single Cycle Execute, which for I/O is Single Character Execute), and (b) the READER NOT READY lamp lights up.

Operator response is to load the tape.  READER NOT READY goes dark, but AUTO and MANUAL remain lit, and the machine remains paused.  Operator hits START, and we're off to the races.

In SIMH, I think we should get the usual PTR EOF message and the command prompt.  Operator attaches a tape and says `go`.  SIMH resumes the RxPT instruction where it left off. (At the beginning, in this case.  And most cases; EOF without EOL is pathological.).

It's also standard MI-015 procedure to run off the end of the tape.  Again, we get READER NOT READY and AUTO+MANUAL.  Operator response is to hit RELEASE, set SW1, and hit START.  MI-015 prints out the hash total.

In SIMH, we get PTR EOF and command prompt.  Operator does something to RELEASE.  A new SIMH command, `release`, or if that's hard or a Bad Idea, `d release 1` or something like that.  Then `go`.  SIMH sees release=1 and does not resume the I/O; instead it starts at the following instruction as if the I/O had previously completed.  Or released, in 1620 jargon.

This approach is also fine for RxTY.  Operator hits ^E.  This gets the `sim>` prompt but does not RELEASE.  Operator sets console switches or whatever he needs to do, says `go`, SIMH resumes the tty input operation, and the user resumes typing and/or hits Return which causes RELEASE-START.  Or the user could cause RELEASE, and then `go`.  Note that a separate RELEASE analogue is not needed for tty anymore, but it does no harm, can be used for a realistic feel, and it was needed anyway for ptr.

**Comment by TomMcBride** [comment]
Dave,

I need to do a bit of testing to make sure I completely understand all of this. In general, with the sole exception of interrupting a read from the typewriter which you/we have been working on, I saw no need to stop the simulator in the middle of an instruction (other than the case of a bad instruction which is equivalent to check stop and no execution could continue without operator intervention). The cases you bring up where a device is not ready through a different light on the subject. For output devices, can this really happen? If a card or tape punch or a printer is not ready, this seems to me the equivalent of not having the device defined to the machine. Perhaps it would be nice to be able, at that point, to attach a file and have the machine rerun the output instruction. In the case of a read to a card reader or tape reader, things are a bit different. This means that we have already reached end of file on the device and the program has issued another read. At this point you are at a simh prompt and can issue a command to reposition the file or detach the file that is finished and attach another and re-execute the instruction. The case of reading a tape with no end marker is a bit of a problem. Would a real machine continue if another tape was loaded in the reader? Just press start?

I don’t have too much time the next couple of days but will work on this.

Tom

**Comment by DaveWise2** [comment]
The paper tape case is from MI-015.  It reads until there's nothing more to read, which requires manual intervention, either RELEASE or load more tape.  Obviously you don't want the latter, because it would change the hash, so we must RELEASE.  We can't just 'go', that will re-execute the RAPT.  We could 'd PC <whatever PC+12 is>' but that's so awkward and unlike physical operation.

I thought about it some more, and I no longer think I/O resume is necessary, just release.  I will work on that and look for cases where it's not sufficient.

**Comment by DaveWise2** [comment]
Here is my proposed fix.
[436_patch_2.diff.txt](https://github.com/simh/simh/files/1005470/436_patch_2.diff.txt)

If you `go` after I/O that broke via ^E or EOF, the I/O resumes where it left off instead of restarting.  RELEASE is accomplished by `d release 1` and `go`.  As the transcript below shows, I get a realistic operator experience for the two problem cases, namely, MI-015, and the 'goof loop'.

```
IBM 1620 simulator V4.0-0 Beta        git commit id: 6a180c3d
sim> cd c:\temp
sim> attach ptr mi015.pt
sim> boot ptr

HALT instruction, PC: 12 (B 00402,00071,2)
sim> attach ptr mi015.pt
sim> g
PTR end of file

End of file, PC: 486 (RAPT 01021)
sim> d sw1 1
sim> d release 1
sim> g

6466670329279453658`3
HALT instruction, PC: 794 (B 00402,03400)
sim> d sw4 1
sim> attach ptr fo005.pt
sim> b ptr

1620 PRE-COMPILER  11/15/61
ENTER SOURCE PROGRAM
THEN PUSH START
HALT instruction, PC: 1460 (B 14860,00000)
sim> attach ptr bad65.pt
sim> g

      DIMENSION FOM(130),IPER(130)
     C  RECORD FIGURE OF MERIT FOR EACH CANDIDATE
      DO 10 I = 65,130
      IPER(I) = I
      FREQ = I
      FREQ = 1.E6 / FREQ
      FINT = FREQ / 95. + 0.5
      ICYC1 = FINT
      FINT = FREQ / 162. + 0.5
      ICYC2 = FINT
      F1 = ICYC1
      F1 = FREQ / F1
      F2 = ICYC2
      F2 = FREQ / F2
      R1 = ((F1 / 95.0) - 1.0) * 100.0
      R2 = ((F2 / 162.0) - 1.0) * 100.0
      FOM(I) = R1 - R2
      IF (R1 - R2) 5,10,10
    5 FOM(I) = R2 - R1
   10 CONTINUE
     C  SORT BY FIGURE OF MERIT, BEST FIRST
     C  1 = NOT SORTED, 2 = SORTED
   20 ISRTD = 2
      DO 25 I = 65,129
      J = I + 1
      IF (FOM(J) - FOM(I)) 22,25,25
   22 T = FOM(I)
      FOM(I) = FOM(J)
      FOM(J) = T
      IT = IPER(I)
      IPER(I) = IPER(J)
      IPER(J) = IT
      ISRTD = 1
   25 CONTINUE
      GOTO (20,30),ISRTD
     C  ELIMINATE INFERIOR ENTRIES - SHORTER PERIOD WITH WORSE FOM
   30 DO 40 I = 65,130
      DO 40 J = I,130
      IF (IPER(J) - IPER(I)) 33,40,40
   33 IPER(J) = 0
   40 CONTINUE
STNO 03

     C PRINT WHATS LEFT
>     c  print whats left
Simulation stopped, PC: 1772 (RATY 16355)
sim> d sw4 0
sim> d release 1
sim> s

Step expired, PC: 1784 (BC4 01736)
sim> s

Step expired, PC: 1796 (BNR 01724,16498)
sim> d sw4 1
sim> g

     C  PRINT WHATS LEFT
      TYPE 90
      DO 50 I = 65,130
      IF (IPER(I)) 50,50,45
   45 FREQ = I
      FREQ = 1.E6 / FREQ
      FINT = FREQ / 95. + 0.5
      ICYC1 = FINT
      FINT = FREQ / 162. + 0.5
      ICYC2 = FINT
      F1 = ICYC1
      F1 = FREQ / F1
      F2 = ICYC2
      F2 = FREQ / F2
      R1 = ((F1 / 95.0) - 1.0) * 100.0
      R2 = ((F2 / 162.0) - 1.0) * 100.0
      TYPE 100, IPER(I),ICYC1,ICYC2,F1,F2,R1,R2,FOM(I)
   50 CONTINUE
      STOP
   90 FORMAT (1HT,4HCYC1,4HCYC2,2HF1,2HF2,2HR1,2HR2,8HMISMATCH)
  100 FORMAT (I3,I3,I3,F6.2,F6.2,F4.2,F4.2,F4.2)
      END

OBJECT PROGRAM DATA TABLE
`03080 STORAGE POSITIONS

PROCESSING COMPLETE

HALT instruction, PC: 9572 (B 01208,00000)
sim>
```

**Comment by TomMcBride** [comment]
Dave,

This seems good to me. I don’t really understand all of it. I have worked on the simulator extensively but worked on the code to simulate instructions and really didn’t look much at the way the terminal was handled or the way commands were processed. I especially find the break-in process hard to understand. Before getting involved with it as a result of your work & questions, I thought it only interrupted between instructions which means that it could not break in to the middle of an I/O operation. Some of my tests have been puzzling. For example, I wrote a program to multiply two 15,000 digit numbers (so it would take long enough to be able to hit ^E before it finished the instruction). I write out a line that says starting the multiply, hit ^E quickly, and expected to have the interrupt at the instruction after the multiply but I don’t see that — it interrupt but shows to be at various other instructions after the multiply. I’m still looking at this to understand it. I see you did not include my extra line of writing the ^ on interrupting a read from the typewriter. I think it is a help. Here is a copy of that section of the code:

/* Read from keyboard */

t_stat tti_read (int8 *c)
{
int32 t;

do {
    t = sim_poll_kbd ();                                /* get character */
    if (stop_cpu == 1) {        /* NEW  from Dave */
        tto_write ('^');        /* tfm  try printing ^ if break */
        return SCPE_STOP;       /* NEW  from Dave */
      }
    } while ((t == SCPE_OK) || (t & SCPE_BREAK));       /* ignore break */
if (t < SCPE_KFLAG)                                     /* error? */
    return t;
*c = t & 0177;                                          /* store character */
return SCPE_OK;
}

If I understand your changes correctly, there is the possibility of interrupting a long I/O operation like reading a long paper tape. Is that true? And your new release switch can be used to terminate it or not. If a read instruction to the card reader or paper tape is issued and the device is not ready, do your changes make it possible to attach a file and continue that instruction as if nothing was ever wrong? I had the sense that this was possible and what you wanted to do.

I’ve seen output from the diff command before and I think I understand what I’m seeing but how do I incorporate those changes into my code so I can try it? Is there some easy way? What is the procedure for getting your changes into i1620? In the past, I just sent them to Bob Supnik and he took care of it. I am registered on GitHub so I can read problems but I have no idea how to actually use it to change code or even if I authorized to do so.

On a different issue, I’ve been working on creating examples for a document I’m working on and I tried changing the break-in character with the command 

SET CONSOLE WRU=26

which I thought was ^Z. It took the command but does not interrupt when I press the key combination. I don’t know what’s going on. I can’t seem to find where this command is processed. It does show my change when I enter SHOW CONSOLE — either I don’t understand what I’m supposed to type or I’m putting in the wrong value or there’s a bug.

Also, while I was working on examples for my document, I put in an instruction with an invalid opcode to show what the error looked like. I was surprised to get a bad display. I used opcode 87 and the display says the opcode is invalid, gives the correct location of the instruction but shows the op code as 08 instead of 87. I’m still looking for the place where this is printed so I can fix it. I’m pretty sure I know what’s wrong. Internally, each location of 1620 storage is represented as 5 bits: F8421. This means that the opcode of 87 is represented as 87 but I think the output is breaking the first character (8) into two 4 bit pieces (0 and 8) and displaying them. It should be taking the lower 4 bits from each of the two storage locations and displaying that instead. There is also the associated question of what will it display if one or both of the locations for the invalid opcode contain a record mark, group mark or numeric blank. It could show exactly what is in storage for the low order bits, namely, A, F, C (hex values).

I’ll just mention one option you may not have noticed. It’s the SET TTY 1DIGIT. This causes the display of flagged data to be done with only one digit instead of 2. So from the console you show in your note, it would show 

> 6466670329279453658L instead of
> 6466670329279453658`3
This is what would have shown on a printer; the typewriter would have printed the 3 with a flag over it. The default is TTY 2DIGIT which prints, as you see, with a ` character in front of a flagged digit. The one digit version is, I think, preferable when you are printing numeric data that should line up in columns.

Tom

**Comment by DaveWise2** [comment]
>For example, I wrote a program to multiply two 15,000 digit numbers (so it would take long enough to be able to hit ^E before it finished the instruction). I write out a line that says starting the multiply, hit ^E quickly, and expected to have the interrupt at the instruction after the multiply but I don’t see that — it interrupt but shows to be at various other instructions after the multiply.

That's working exactly as designed, you can't interrupt an instruction in the middle, only on the boundaries between instructions.  That's equivalent to the 1620 STOP/SIE button.  Single Instruction Execution, vs the STOP/SCE button which interrupts on machine cycle boundaries for non-I/O, and on character boundaries for I/O.  There's no point to machine-cycle SCE for SIMH, but the I/O interrupt I'm shooting for will be equivalent to 1620 I/O SCE, and it's essential for some nonstandard but common procedures.

>writing the ^ on interrupting a read from the typewriter

I do like it, and I want it to go in.  However, tto_write() won't print it; with cpu_stop == 1, it will exit without doing it.  We have to use sim_putchar() directly, or save, clear, and restore cpu_stop around the call - ugh.

>If I understand your changes correctly, there is the possibility of interrupting a long I/O operation like reading a long paper tape. Is that true?

Yes, but the real place it's useful is to get out of a read loop with the same operator experience as on a real 1620.  When a program reads off the end of a tape (some have to, some don't know any better), the big red RDR NOT READY lamp comes on, AUTOMATIC stays on, MANUAL comes on (this combination indicates an intervention-required pause), you load up another section of tape, RDR NOT READY goes out, you hit START, and the CPU resumes where it left off.  Or, you hit RELEASE, and the read is over.  And then maybe you flip a switch to tell the program to stop looping on read and do the next thing.
Currently, you can fake it in SIMH with `d pc <pc + 12>` where you have to calculate pc + 12 - an artificial and unnecessary hassle.

Now understand, I'd like a `release` command, but I don't see that happening.  `d release 1` isn't too bad.

>do your changes make it possible to attach a file and continue that instruction as if nothing was ever wrong?

Exactly.  And it's a common event in a real 1620 shop.

>I’ve seen output from the diff command before and I think I understand what I’m seeing but how do I incorporate those changes into my code so I can try it?

I haven't tried it in tortoiseGit, but tortoiseSVN (with an intentionally similar UI) has an "Apply Patch" command.

>What is the procedure for getting your changes into i1620? In the past, I just sent them to Bob Supnik and he took care of it.

I hope to hear from Mark P about this.  I wouldn't mind being able to "push" my local repository upstream, but that's a lot of power in the hands of a newbie.  I need training wheels. :)

<WRU setting, and bad display of invalid opcode> could very well be bugs.  The former is common to all targets, and Mark and Bob should know right off the bat.  I don't know.

Thanks for mentioning SET TTY 1DIGIT.  I didn't know about it, but it makes perfect sense.  As you say, it's what a 1443 would print.  Or a 407.  With the HASH TOTAL program, there's just the one line - and different distribution memos formatted them differently - so it doesn't matter.

**Comment by TomMcBride** [comment]
Dave,

If I understand your comments correctly, printing ^ after entering ^E may be hard if done for all cases but it’s easy, as my code does, to print it on interrupting a read from the typewriter. I thought that alone would be worthwhile. I do think it would be nice if it printed everywhere but I’m not sure it’s worth the trouble. I don’t understand your comment about tto_write() exiting with cpu_stop == 1. There is no test of cpu_stop in tto_write(). It’s printing my ^ just fine. Now, trying to call it from somewhere else when it has not been initialized or done whatever setup it needs for writing would likely be a problem.

My tests SEEM to show that ^E during my big multiply is not stopping immediately after that instruction finishes as I expected. I understand there will be no interrupt during the instruction. I will improve my test to be sure. I can also move my test case to a Raspberry Pi zero which is much slower than my Mac so the multiply will take longer. 

Have you tested your changes using the card reader? If the reader is empty (all the cards have been read) and a new read is executed, can you then attach a different file to the reader and continue? That would certainly be nice.

The incorrect display when a bad opcode is encountered is definitely a bug. I’m not sure about the SET CONSOLE WRU=xxx but it sure seems to be a bug.

The SET TTY 1DIGIT was my addition; the variable length displays were driving me nuts trying to look at columns of numeric data. I also added the code to accept ]JKLMNOPQR and ]jklmnopqr characters on numeric read as equivalent to the flagged digits 0-9 — as an addition to the original  convention of `1 or ~1 for a flagged 1, etc. on input. To me it made sense since all 1620 users were familiar with the equivalence between the letters and flagged digits. And, of course, if you created a file to be read by the card reader, it had to have these characters for the flagged digits. That’s what one sees viewing a punched card deck in an editor.

Tom

**Comment by DaveWise2** [comment]
>I don’t understand your comment about tto_write() exiting with cpu_stop == 1. There is no test of cpu_stop in tto_write()

I added one.  But sim_putchar() works fine.

>My tests SEEM to show that ^E during my big multiply is not stopping immediately after that instruction finishes as I expected. I understand there will be no interrupt during the instruction.

Sorry, I misunderstood.  Good luck; timing problems can be hard to pin down.

>Have you tested your changes using the card reader?

Not yet.  On EOF, i1620_cd.c returns STOP_NOCD.  I added that to my resume test.  It should feel just like a 1622 - running out of cards, or START or LOAD without any cards, lights the RDR NOT READY and MANUAL lamps, and AUTOMATIC stays lit.  You dump some in the hopper, RDR NOT READY goes dark, you hit START, and off you go.

Letters for flagged digits was actually introduced with the 1622; since I don't have one, I still haven't memorized them even after all these years.

Dave

**Comment by archivisth** [comment]
There have been a lot of suggestions, comments, and so on. I'd like to reiterate what I said before and what I plan on doing.

There are three separate problems.

1. ^E doesn't interrupt keyboard input on some operating systems.

This is easily fixed by testing stop_cpu in the keyboard polling loop. The 1401 already does this.

2. The Model II doesn't implement backspace.

This requires recognizing backspace as a legitimate character on a Model II and letting it propagate back up to RA and RN for the TTY, because that's where the buffer pointer is kept.

3. The PC is incorrect on many/most error conditions.

This was a deliberate decision, so that the error message display pointed at the instruction in question, rather than the next instruction. It means that ^E from a R*TTY will repeat the instruction rather than going on to the next one. This is incorrect behavior. The PC should, in theory, always be advanced as soon as instruction decode is complete.

There are two ways to deal with problem #3.

3a. Always advance the PC, as the hardware does. That means if a paper tape or card IO fails because there's no attached file, the PC must be backed up by hand to re-execute.
3b. Advance the PC selectively, based on the stop code. In particular, SCPE_STOP must advance the PC. I haven't look at the 39 unique 1620 error codes, and all of the other SimH error codes.

The fix I am looking at is 3b: commit the instruction on certain select stop codes, like SCPE_STOP; but always store both the 'committed' PC and the actual value of the PC at the time of the stop. A new command, 'RELEASE', would would set the committed PC to the actual PC. If they were the same, it would have no effect, obviously.

I have not looked through the various enhancement suggestions, but I'm not very interested in replicating the 'feel' of a real 1620. For example, being able to interrupt out of non-keyboard IO doesn't make sense to me, because simulator IO is so much faster than the real world. A paper tape or card read will run as fast as the host operating system can supply characters. The simulator won't allow "runaway" IO, so the largest transfer (60,000 digits) will be over in a flash. Likewise, interrupting out of MUL and DIV doesn't make much sense, because the instructions are destructive; they can neither be restarted nor continued.

**Comment by TomMcBride** [comment]
Bob,

I think your ideas are good. The problems of interrupting out of an I/O operation, at least to my mind, are of two kinds. One is the ability to interrupt a read to the typewriter so that some simulator commands can be issued and the program restarted for purposes of correcting a typing error. The second case is the ability to get control when a read instruction is for a tape or card reader but there is no data to read. The idea is to be able to issue simulator commands to attach a file to the device and resume the program. 

I understand and agree with you about not needing to simulate the actual “feel” of the 1620 so I’m not concerned that the operating procedures are a bit different. I did like my change which prints a ^ on the line when a typewriter read was interrupted with ^E. There is a similar issue with pressing the enter key to end a read from the typewriter in that a real 1620 typed a special character (an RS ligature) to show on the page how much was typed. This made it possible to see how many blanks we’re entered at the end of an input. Clearly, it’s not possible to type that exact character but what do you think of tying something else such as a > for example?

I did not mean to imply that in my comments that individual instructions should be interruptible. It’s important to be able to interrupt execution when a program loop is suspected, issue debugging commands, and then resume execution. My observation (which I have not confirmed to my satisfaction) was that it did not seem to interrupt after the instruction that was running at the time. I was using a very lengthy instruction so I could be sure it was running when I entered ^E. It may be that I’m not fast enough. I’m still looking at this.

Tom McBride

**Comment by DaveWise2** [comment]
Tom beat me to it.

Thanks, Bob, I very much prefer a `release` command to my `d release 1` flag hack.  I just didn't know how to add a 1620-specific command, and I wasn't sure you or Mark would approve.

I wasn't pushing for interrupting file-attached IO, only for covering the cases where it's necessary to release out of it, attach a new file, or manipulate switches.  But as this facilitates running all valid programs, it does come closer to the real 1620 feel, even if you're not aiming for it as such.

Simulating machine cycles is out of scope and not worth it.  And Chap Harrison already did it.  But interruptible IO does echo the "stop" aspect of the 1620 STOP/SCE button.

>The fix I am looking at is 3b: commit the instruction on certain select stop codes, like SCPE_STOP

Unless I've missed some subtleties, resuming streaming IO (pt, cd, tty) where it left off isn't hard.  (And it makes commit irrelevant.)  It's not necessary, strictly speaking, but isn't it tasty?  Would you mind critiquing my patch offline?

> I was using a very lengthy instruction so I could be sure it was running when I entered ^E. It may be that I’m not fast enough. I’m still looking at this.

Tom, the I1620 throttling mechanism sleeps between instructions; it does not slow down the instructions themselves.  As far as timing is concerned, huge multiply == NOP.  (Mark recently did add console output throttling, `SET CONSOLE SPEED=nnn`, where nnn is a baud rate.)

**Comment by markpizz** [comment]
> I understand and agree with you about not needing to simulate the actual “feel” of the 1620 so I’m not concerned that the operating procedures are a bit different. I did like my change which prints a ^ on the line when a typewriter read was interrupted with ^E. There is a similar issue with pressing the enter key to end a read from the typewriter in that a real 1620 typed a special character (an RS ligature) to show on the page how much was typed. This made it possible to see how many blanks we’re entered at the end of an input. Clearly, it’s not possible to type that exact character but what do you think of tying something else such as a > for example?

Tom, the simh framework starts out with two separate activities happening on the session which starts the simulator.  These are:
 1) the interactions with the simulator framework
 2) the I/O which happens on the console port of the system being simulated.

The simh framework only starts in this mode, you can completely separate these two independent things with: 

sim> SET CONSOLE TELNET=tcp-port

and then starting a telnet session to localhost:tcp-port.

^E is detected only in the context of the interactions which control the simulator.  It likely means nothing in a telnet session (as it may have with the original system being simulated).

Attempting to nicely accommodate the dual purpose startup condition of the simh framework really shouldn't be a goal of anything in the simulator itself.

**Comment by archivisth** [comment]
The TTY changes are done and seem to work; now on to "What should the PC be at a stop?"

The PC can only be in one of three states:
- The original PC.
- PC + 12, after successful execution of an instruction.
- Branch PC, after successful execution of a branch.

If an error is detected during instruction parsing (e.g., invalid digit in P or Q addresses), the internal PC has not been changed.

If an error is detected after parsing but before successful completion (e.g., invalid digit in a field), the internal PC has been advanced but not committed.

If the instruction completes successfully, the new PC (PC +12 or branch PC) is committed.

This works well in most cases, EXCEPT FOR:

^E out of RATY or RNTY. In general, ^E should commit any change to the PC.

End of file, no card, or other IO error. If data runs out before an instruction finishes, any changes to the PC should be committed. The instruction cannot be repeated, because memory state has been changed.

So let's look at some scenarios:

1. "Goof" loop. ^E will advance to the next instruction. The RATY or RNTY is effectively terminated.
2. No input or output file file (SCPE_UNATT). The PC is not advanced. The user can attach a file and enter GO to try the instruction again or type RELEASE to go to the next instruction.
3. End of file (SCPE_IOERR or STOP_NOCD). The PC is advanced. If the user enters GO after the error, the next instruction is executed.

I'm on the fence about "LAST CARD". Unlike paper tape reader end of file, if LAST CARD is detected, nothing in memory has been changed, because each card reader RA or RN reads exactly one card. It would be possible to load another deck (attach another file) and continue, re-executing the current instruction.

**Comment by DaveWise2** [comment]
Thanks, Bob.  I'm looking forward to studying your changes.

We may need to ^E out of WxTY/DNTY, otherwise we can be trapped for a while during baud-reduced redirected console output.

Goof loop.  If ^E advances, then it's effectively a RELEASE.  It shouldn't, it should resume wherever it left off.  It's not unusual to type, realize you need to flip a switch, then finish typing.  I just sent Mark a change rollup [mark_2.zip.txt](https://github.com/simh/simh/files/1012113/mark_2.zip.txt) that does this.  I hope you can look it over.

EOF/NOCD.  A PT file that doesn't end with EOL is pathological, so EOF means the program wants to read another record and it isn't there.  Just like NOCD.  This is invariably a program that expects you to RELEASE, flip a switch, then GO on to the next phase.  EOF - at least at the beginning of read - should not advance.

**Comment by TomMcBride** [comment]
Bob,

This all looks good to me. I don’t think you need to make any changes with regard to the LAST CARD indicator. This indicator comes on AFTER the last card has been read. No cards remain in the reader. I believe the last card indicator is reset when tested but I’m not absolutely certain.  I would hope that if a new file is attached to the reader, the last card indicator would also be reset. My memory of how this worked is not perfect. If the reader is empty and the program tries to read another card, a light comes on the console indicating the reader is selected and the machine hangs until the reader is made ready. I’m unable to find all of these details in any of the IBM books. I hope Dave knows or can figure them out.

A common program sequence was: (1) read a card, (2) do some processing, (3) BNLC (branch not last card) back to (1). If the stop key on the reader was pressed (e.g., so that more cards could be loaded), the machine would hang on the next read until the start key on the reader was pressed.

Tom

**Comment by DaveWise2** [comment]
Testing indicator 9 (LAST CARD) turns it off.  So does RESET.  See A26-4500-2 page 47.

**Comment by DaveWise2** [comment]
Bob, I withdraw my assertion that NOCD should not advance.  I have no opinion one way or the other.  I had forgotten about the LAST CARD indicator.  Unlike paper tape, the program has a way to tell when it's about to run out of cards.  Any program that ignores that is badly behaved.  For paper tape, there's no way to tell, so all the poor program can do is read until we lean in and flip a switch.

**Comment by DaveWise2** [comment]
Something I keep forgetting to mention.  RESET should get you out of I/O if the CPU was stalled on EOF etc.

**Comment by archivisth** [comment]
Changes have been committed for:

1. Implementation of backspace on Model II.
2. ^E interrupting type-ins on all operating systems.
3. Instructions being marked as completed if interrupted or if an IO error occurs.
4. New command, SET CPU RELEASE, implemented to advance beyond an instruction that throws an error.

The list of error stops that commit or repeat an instruction is easily adjusted - it's driven off a table of error stop codes.

The original issue can be (and should be) closed. Many of the issues raised during the discussion relate to the divergence between the simulator's model of IO and the real machine's. This can't be patched through or around. Rather, the IO design needs to be redone to more closely mimic how the actual hardware works. The simulator has to become bimodal - it is either executing instructions, or it is performing IO - using the same basic loop structure and mechanics. This will make <all> IO interruptible and continuable, unless RELEASEd.

Such a radical redesign is going to take time and testing, so please be patient.

**Comment by archivisth** [comment]
The design for 'more accurate' IO requires fewer changes that I thought. Basically, at IO initiation, the device needs to remember the IO type (RA, RN, WA, WN, DN), the maximum transfer count (device and operation specific); the starting/current address is always in PAR. It then schedules the first transfer, sets an 'IO in progress' flag, and returns, retiring the IO instruction.

If 'IO in progress' is set, the CPU advances the event timer and calls the event scheduler; instruction execution is skipped. When the IO device event routine is called, it transfer a digit/character to the PAR address and advances PAR, with one of three outcomes:

- More to do; schedule next transfer and return OK.
- Error: schedule next transfer and return error code. This will stop simulation and return to the console. However, 'IO in progress' is still set, so the IO operation will resume if the user just continues.
- Successful completion: clear 'IO in progress', do not schedule another transfer, and return OK.

The release command (and reset) simply clear 'IO in progress'. There's no need to cancel the next transfer. All the IO event routines will check 'IO in progress' first. If it is clear, then they simply exit. (The keyboard event routine will always be scheduled, to handle ^E, but it will ignore any input if 'IO in progress' is not set.)

With this change, IO behaves the way a real 1620 behaves: IO stops execution until either the IO operation completes successfully or RELEASE is pressed (for the simulator, SET CPU RELEASE).

I also think the simulator should handle the PC the way the real system does. Once decode is complete, the simulator will advance the PC by 12. To help with errors that occur during instruction execution, the stop message routine will print out additional information pointing to the erroneous instruction on any sort of instruction check stop (e.g., bad digit). The error routine should also flag whether IO is in progress.

I don't think it's worth changing all of the IO devices to this model, only the ones where interruptability and restartability matter - the typewriter in and out, and the paper tape reader and punch. The line printer is limited to a maximum of 120/144 characters, the card reader to one card, and the disk to one track. Even the card punch in DN mode, which can punch up to 251 cards, completes quickly, because it always outputs to a file. I'm open to feedback on this, but leaving the reader/punch, printer, and disk unchanged minimizes the amount of testing needed.

**Comment by DaveWise2** [comment]
Thanks, Bob, I'm looking forward to studying and testing your latest changes.  I think there was a problem with your May 20 changeset.  I imagine you've already fixed it or nullified it via further changes, but for the record: Since SCPE_STOP is in commit_pc[], `^E` will set actual_PC = saved_PC = PC+12, and a subsequent `set cpu release` will do nothing, the IO already having been effectively released.

**Comment by markpizz** [comment]
Sounds reasonable, with one question:

> If 'IO in progress' is set, the CPU advances the event timer and calls the event scheduler; instruction execution is skipped. When the IO device event routine is called, it transfer a digit/character to the PAR address and advances PAR, with one of three outcomes:

If 'IO in progress' is set, if sim_interval <= 0 then call sim_process_event(), else sim_idle(0, 1).

**Comment by archivisth** [comment]
This issue can be closed with the current commit, which implements deferred IO for the teletype and paper tape equipment.

**Comment by DaveWise2** [comment]
Thanks, Bob, I'm looking forward to trying it out, and studying your changes.  I have been dealing with an emergency, but I think I can resume here soon.

**Comment by DaveWise2** [comment]
Looks good in casual testing.  I'll post here if I find anything.
